src/realm/object-store/c_api/socket_provider.cpp

```diff
@@ -0,0 +1,255 @@
+};
+
+struct CAPISyncSocketProvider : sync::SyncSocketProvider {
+    realm_userdata_t m_userdata = nullptr;
+    realm_free_userdata_func_t m_free = nullptr;
+    realm_sync_socket_post_func_t m_post = nullptr;
+    realm_sync_socket_create_timer_func_t m_timer_create = nullptr;
+    realm_sync_socket_timer_canceled_func_t m_timer_cancel = nullptr;
+    realm_sync_socket_timer_free_func_t m_timer_free = nullptr;
+    realm_sync_socket_connect_func_t m_websocket_connect = nullptr;
+    realm_sync_socket_websocket_async_write_func_t m_websocket_async_write = nullptr;
+    realm_sync_socket_websocket_free_func_t m_websocket_free = nullptr;
+
+    CAPISyncSocketProvider() = default;
+    CAPISyncSocketProvider(CAPISyncSocketProvider&& other)
+        : m_userdata(std::exchange(other.m_userdata, nullptr))
+        , m_free(std::exchange(other.m_free, nullptr))
+        , m_post(std::exchange(other.m_post, nullptr))
+        , m_timer_create(std::exchange(other.m_timer_create, nullptr))
+        , m_timer_cancel(std::exchange(other.m_timer_cancel, nullptr))
+        , m_timer_free(std::exchange(other.m_timer_free, nullptr))
+        , m_websocket_connect(std::exchange(other.m_websocket_connect, nullptr))
+        , m_websocket_async_write(std::exchange(other.m_websocket_async_write, nullptr))
+        , m_websocket_free(std::exchange(other.m_websocket_free, nullptr))
+    {
+        REALM_ASSERT(m_free);
+        REALM_ASSERT(m_post);
+        REALM_ASSERT(m_timer_create);
+        REALM_ASSERT(m_timer_cancel);
+        REALM_ASSERT(m_timer_free);
+        REALM_ASSERT(m_websocket_connect);
+        REALM_ASSERT(m_websocket_async_write);
+        REALM_ASSERT(m_websocket_free);
+    }
+
+    ~CAPISyncSocketProvider()
+    {
+        m_free(m_userdata);
+    }
+
+    std::unique_ptr<sync::WebSocketInterface> connect(sync::WebSocketObserver* observer,
+                                                      sync::WebSocketEndpoint&& endpoint) final
+    {
+        auto capi_observer = std::make_shared<CAPIWebSocketObserver>(*observer);
+        return std::make_unique<CAPIWebSocket>(m_userdata, m_websocket_connect, m_websocket_async_write,
+                                               m_websocket_free, new realm_websocket_observer_t(capi_observer),
+                                               std::move(endpoint));
+    }
+
+    void post(FunctionHandler&& handler) final
+    {
+        auto shared_handler = std::make_shared<FunctionHandler>(std::move(handler));
+        m_post(m_userdata, new realm_sync_socket_callback_t(std::move(shared_handler)));
+    }
+
+    SyncTimer create_timer(std::chrono::milliseconds delay, FunctionHandler&& handler) final
+    {
+        auto shared_handler = std::make_shared<FunctionHandler>(std::move(handler));
+        return std::make_unique<CAPITimer>(m_userdata, delay.count(),
+                                           new realm_sync_socket_callback_t(std::move(shared_handler)),
```

Comment:
same question about the lifetime of this callback.

Reply:
Since this is a timer, it will eventually call `post` and `realm_sync_socket_callback_complete`, so the pointer will be free'd (as per above update)